Add documentation

justjoheinz opened the pull request:
Fix errors and add some smaller hints, like using modifyF with a Future (common use case)

justjoheinz commented:
I plan to add some more documentation. Please let me know if you prefer one commit or smaller ones. I prefer smaller ones, as I think less documentation is better than none at all. Also please review - part of the excercise for me is to get a better understanding of optics.

julien-truffaut commented on docs/src/main/tut/iso.md:
```diff
@@ -0,0 +1,34 @@
+
+# Iso
+
+An `Iso` is an Optic which converts elements of type `A` into elements of type
```
We generally use type `S` and `A`. It doesn't really matter but I suppose it is better to be consistent.

justjoheinz replied:
ok

julien-truffaut commented on docs/src/main/tut/iso.md:
````diff
@@ -0,0 +1,34 @@
+Consider these two case classes:
+
+```tut:silent
+case class Person(name: String, firstName : String)
````
This example might be more illustrative if the fields in `Person` are of different type e.g. `String` and `Int`

justjoheinz replied:
ok

julien-truffaut commented on docs/src/main/tut/iso.md:
````diff
@@ -0,0 +1,34 @@
+* reverseGet : Pers => Person
+
+```tut:silent
+import monocle._
````
maybe we could use a more precise import, e.g. `monocle.Iso`

julien-truffaut commented on docs/src/main/tut/iso.md:
````diff
@@ -0,0 +1,34 @@
+
+```tut:silent
+import monocle._
+val personPrism = Iso[Person, Pers]((p: Person) => Pers(p.name, p.firstName))((p: Pers) => Person(p.n, p.f))
````
if you want to make this example shorter you can either use:
`Iso[Person, Pers](p => Pers(p.name, p.firstName))(p => Person(p.n, p.f))`
or
Iso((p: Person) => Pers(p.name, p.firstName))((p: Pers) => Person(p.n, p.f))

justjoheinz replied:
In documentation I like to see the types even if they can be inferred as I believe it helps to connect the dots as I cannot get them tooltipped from the screen.

julien-truffaut replied:
it makes sense, maybe we can show a shorter version in other example or as a note.

julien-truffaut replied:
also the name of the `Iso` contains `Prism`

julien-truffaut commented on docs/src/main/tut/iso.md:
```diff
@@ -0,0 +1,34 @@
+# Iso
+
+An `Iso` is an Optic which converts elements of type `A` into elements of type
+`B` without loss.
```
later we might want to mention that `Iso` comes from Isomorphism and point to wikipedia or other resources

julien-truffaut commented on docs/src/main/tut/iso.md:
```diff
@@ -0,0 +1,34 @@
+
+and thereby create a lossless conversion between these two types. Other usages include for example the transformation of different types of physical units.
+
+It is important to understand that the transformation between type `A` and `B` works for any type `A` and `B` and that the functions `get` and `reverseGet` are true inverses to each other.
```
:+1: to mention that `get` and `reverseGet` are inverse of each other, we could also give the law that comes from this. However, I find the first sentence ("It is important") a bit confusing.

julien-truffaut commented on docs/src/main/tut/lens.md:
````diff
@@ -50,15 +50,32 @@ We can push push the idea even further, with `modifyF` we can update the target
 _streetNumber.modifyF(neighbors)(Address(135, "High Street"))
 ```
 
+This would work with any kind of `Functor` and is especially useful in conjunction with asynchronous APIs, where one has the task to update a deeply nested structure (see Lens Composition) with the result of an asynchronous computation:
````
I would say any `Functor` or any type `F` that has an instance of `Functor`

julien-truffaut commented on docs/src/site/optics.md:
```diff
@@ -6,5 +6,12 @@ section: "optics"
 
 # Optics
 
+## Convert between two types
```
I am not sure we should make that kind of categorisation here. Maybe it would be a good place to put the class diagram that shows relationship between optics.

justjoheinz replied:
I would like to rework that diagram eventually. It confused me totally when I tried to learn about monocle, because I constantly thought it was talking about inheritance and it took me awhile to realize that it is talking about transformations (I am UML damaged to certain degree). Thus I never understood why my lenses did not compose.

julien-truffaut replied:
At some point (0.5 or 0.6), optics were actually inheriting from each other. We removed it when we started to encode them as case class, nowadays I am not sure if we should use or not inheritance. I

jbgi replied:
Having access to the Prism (weaker) api in Iso would be confusing, better not use inheritance.

julien-truffaut commented:
Thanks @justjoheinz it is very helpful! I made many small comments, you don't have to take care of all of them, other people including yourself might disagree. Also I am not a native speaker, so I am definitevely not the best person to correct formulation.

justjoheinz commented:
I just incorporated some changes, and missed some you were obviously just making. I also added a "tut" target to travis.yml - the docs were already containing errors, and I hope we agree, that this should not happen.

julien-truffaut commented on .travis.yml:
```diff
@@ -13,7 +13,9 @@ notifications:
   email:
   - truffaut.julien@gmail.com
-script: sbt ++$TRAVIS_SCALA_VERSION test
+script:
```
great I didn't notice

julien-truffaut commented:
@justjoheinz if you can rebase this PR, I will be happy to merge it.

julien-truffaut commented:
It is weird, it seems that they are still some conflict but we haven't merged anything in the meantime.

justjoheinz commented:
I think I mucked it up, when rebasing, will give it another try tommorow, had been a busy day.

julien-truffaut commented:
np, please tell me if I can help :)

julien-truffaut commented:
@justjoheinz I propose that we close this PR until you or anyone else get sometimes to work on it.

justjoheinz commented:
Ok, maybe the easiest for me would be to close this, and just reopen a new PR based on the current state. Sorry - have been away and busy.

justjoheinz commented:
I just opened another branch created from a clean slate.
